### PR TITLE
Add Single, SingleOrDefault, SingleAsync, SingleOrDefaultAsync

### LINKED ISSUE #94: Operator: Single
[Link to Rx.NET implementation](https://github.com/dotnet/reactive/blob/main/AsyncRx.NET/System.Reactive.Async.Linq/System/Reactive/Linq/Operators/Single.cs)

### PR DESCRIPTION
**Pull request by ronnygunawan**
Resolves #94

### CODE REVIEW

File: Reaqtive/Core/Reaqtive.Linq/Reaqtive/Operators/SingleAsync.cs [hunk 1]
```diff
@@ -0,0 +1,125 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT License.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Diagnostics;
+
+namespace Reaqtive.Linq.Reaqtive.Operators
```
Comment (bartdesmet) [inline Reaqtive/Core/Reaqtive.Linq/Reaqtive/Operators/SingleAsync.cs:8]: Abbreviate the namespace to be consistent with all the other operators. I believe VS is trying to be a bit too smart in inferring namespaces from folders here.

File: Reaqtive/Core/Reaqtive.Linq/Reaqtive/Subscribable.cs [hunk 2]
```diff
@@ -5,6 +5,7 @@
 using System;
 using System.Collections.Generic;
 
+using Reaqtive.Linq.Reaqtive.Operators;
```
Comment (bartdesmet) [inline Reaqtive/Core/Reaqtive.Linq/Reaqtive/Subscribable.cs:8]: Clean up (see previous remark).

File: Reaqtor/Samples/Shebang/Reaqtor.Shebang.Service/Linq/QueryOperators.cs [hunk 3]
```diff
@@ -140,6 +140,10 @@ public static async Task DefineAsync(ReactiveClientContext ctx, CancellationToke
...
             await ctx.DefineObservableAsync(new Uri("rx://observable/sequenceequal"), (IAsyncReactiveQbservable<TSource> left, IAsyncReactiveQbservable<TSource> right) => Subscribable.SequenceEqual<TSource>(left.AsSubscribable(), right.AsSubscribable()).AsAsyncQbservable(), null, token).ConfigureAwait(false);
             await ctx.DefineObservableAsync(new Uri("rx://observable/sequenceequal/comparer"), (IAsyncReactiveQbservable<TSource> left, IAsyncReactiveQbservable<TSource> right, IEqualityComparer<TSource> comparer) => Subscribable.SequenceEqual<TSource>(left.AsSubscribable(), right.AsSubscribable(), comparer).AsAsyncQbservable(), null, token).ConfigureAwait(false);
+            await ctx.DefineObservableAsync(new Uri("rx://observable/single"), (IAsyncReactiveQbservable<TSource> source) => Subscribable.SingleAsync<TSource>(source.AsSubscribable()).AsAsyncQbservable(), null, token).ConfigureAwait(false);
```
Comment (bartdesmet) [inline Reaqtor/Samples/Shebang/Reaqtor.Shebang.Service/Linq/QueryOperators.cs:143]: Also add the matching `Undefine` operations below.

**Review by bartdesmet — CHANGES_REQUESTED** [review]
Looks good in general; just a few cosmetic remarks.

Just point out that for new operators, it's always important to make sure that the glitching tests in remoting pass. The tests can be lengthy but can catch subtle issues in state persistence.

File: Reaqtor/Samples/Remoting/Tests.Reaqtor.Remoting.Glitching/Operators/SingleAsync.Declarative.cs [hunk 4]
```diff
@@ -0,0 +1,237 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
```
Comment (bartdesmet) [inline Reaqtor/Samples/Remoting/Tests.Reaqtor.Remoting.Glitching/Operators/SingleAsync.Declarative.cs:1]: Something's a little off here. I'd expect to see this file under `Tests.Reaqtive.Linq\Operators` instead, and then have a link in the `csproj` for `Tests.Reaqtor.Remoting.Glitching`, see https://github.com/reaqtive/reaqtor/blob/main/Reaqtor/Samples/Remoting/Tests.Reaqtor.Remoting.Glitching/Tests.Reaqtor.Remoting.Glitching.csproj.

**Review by bartdesmet — CHANGES_REQUESTED** [review]

File: Reaqtor/Samples/Remoting/Tests.Reaqtor.Remoting.Glitching/Operators/SingleAsync.Declarative.cs (hunk 4 above)
Comment (ronnygunawan) [reply Reaqtor/Samples/Remoting/Tests.Reaqtor.Remoting.Glitching/Operators/SingleAsync.Declarative.cs]: I didn't notice the linking. My bad.

**Review by bartdesmet — APPROVED** [review]
LGTM. It's great to see the effort on extending the operator set.

There are some future optimization opportunities which also apply to related operators such as `First*` with regards to specializations for the four overloads, to reduce the object layout size for the subscribable (which can be kept alive by the Reaqtor query engine). This can be considered separately down the road, combined with efforts to build up a new modern perf suite (e.g. based on BenchmarkDotNet, which wasn't yet in our tools arsenal back when we started work on Reaqtor around 2013, so we used some proprietary tools). There are some tangential aspects to this, related to the way queries are hosted in the query engine, which we can get a fresh look on as well (I'll open issues on such considerations later on).

For now, let's focus on more operator surface parity with classic Rx.

**Pull request merged**